### PR TITLE
8294238: ZGC: Move CLD claimed mark clearing

### PR DESCRIPTION
**Pull request by stefank**
When we claim CLDs during object iteration, we must make sure to have a cleared set of claim bits. Today we ensure this by clearing the bits before object iteration starts. Most GCs perform this clearing during a stop-the-world pause, before the actual GC marking starts.

ZGC, however, performs the clearing concurrently. This requires us to be very careful and never start following object references before the clearing has completed.

In the Generational ZGC repository, we changed it so that the code that performs the object iteration cleans up and clears these bits after itself. This has the effect that when marking starts, we know that the claimed bits have been cleared.

I'd like to change the single-generation ZGC to do the same.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294238](https://bugs.openjdk.org/browse/JDK-8294238): ZGC: Move CLD claimed mark clearing

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [fd2187ac](https://git.openjdk.org/jdk/pull/10591/files/fd2187acd0b4cc74bc2b47ad944d10dde7d6d277)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [d7fdc8e3](https://git.openjdk.org/jdk/pull/10591/files/d7fdc8e3fd82d8ddc2807191fad5abeb1e686f94)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [d7fdc8e3](https://git.openjdk.org/jdk/pull/10591/files/d7fdc8e3fd82d8ddc2807191fad5abeb1e686f94)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10591/head:pull/10591` \
`$ git checkout pull/10591`

Update a local copy of the PR: \
`$ git checkout pull/10591` \
`$ git pull https://git.openjdk.org/jdk pull/10591/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10591`

View PR using the GUI difftool: \
`$ git pr show -t 10591`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10591.diff">https://git.openjdk.org/jdk/pull/10591.diff</a>

</details>

### CODE REVIEW

**Review by coleenp — APPROVED** [review]
CLD bits ok with me.  The ClassLoaderData::verify_not_claimed should probably be in an #ifdef ASSERT conditional.

**Comment by stefank** [comment]
Thanks, Coleen. I've updated the patch with your suggestion.

File: src/hotspot/share/classfile/classLoaderDataGraph.hpp [hunk 1]
```diff
@@ -66,6 +66,7 @@ class ClassLoaderDataGraph : public AllStatic {
   static void purge(bool at_safepoint);
   static void clear_claimed_marks();
   static void clear_claimed_marks(int claim);
+  static void verify_claimed_marks_cleared(int claim);
```
Comment (tschatzl) [inline src/hotspot/share/classfile/classLoaderDataGraph.hpp:69]: Is there a reason to not use `NOT_DEBUG_RETURN` for this method like for `ClassLoaderData::verify_not_claimed()`? I would prefer this to be consistent.

**Review by tschatzl — APPROVED** [review]

File: src/hotspot/share/classfile/classLoaderDataGraph.hpp (hunk 1 above)
Comment (stefank) [reply src/hotspot/share/classfile/classLoaderDataGraph.hpp:69]: I just think the NOT_DEBUG_RETURN / ASSERT macros pollute the code and makes it ugly to read, so I opted to not do it. This function is absolutely not performance critical, so I chose this approach.

I can be compelled to change it, if Reviewers prefer to change the code to use NOT_DEBUG_RETURN .

**Review by fisk — APPROVED** [review]
Looks good.

File: src/hotspot/share/classfile/classLoaderDataGraph.hpp (hunk 1 above)
Comment (tschatzl) [reply src/hotspot/share/classfile/classLoaderDataGraph.hpp:69]: I am fine with the change as is, just noting that this change uses it in one place and does not in the other, which seemed inconsistent to me.

**Comment by stefank** [comment]
Thanks for the reviews!
/integrate

**Pull request closed without merging**